src/pytest_html/html_report.py

```diff
@@ -0,0 +1,342 @@
+                        html.tr(cells),
+                        html.tr(
+                            [
+                                html.th(
+                                    "No results found. Try to check the filters",
+                                    colspan=len(cells),
+                                )
+                            ],
+                            id="not-found-message",
+                            hidden="true",
+                        ),
+                        id="results-table-head",
+                    ),
+                    self.test_logs,
+                ],
+                id="results-table",
+            ),
+        ]
+
+        with open(
+            os.path.join(os.path.dirname(__file__), "resources", "main.js")
+        ) as main_js_fp:
+            main_js = main_js_fp.read()
+
+        body = html.body(
+            html.script(raw(main_js)),
+            html.h1(self.title),
+            html.p(
+                "Report generated on {} at {} by ".format(
+                    generated.strftime("%d-%b-%Y"), generated.strftime("%H:%M:%S")
+                ),
+                html.a("pytest-html", href=__pypi_url__),
+                f" v{__version__}",
+            ),
+            onLoad="init()",
+        )
+
+        body.extend(self._generate_environment(session.config))
+
+        summary_prefix, summary_postfix = [], []
+        session.config.hook.pytest_html_results_summary(
+            prefix=summary_prefix, summary=summary, postfix=summary_postfix
+        )
+        body.extend([html.h2("Summary")] + summary_prefix + summary + summary_postfix)
+
+        body.extend(results)
+
+        doc = html.html(head, body)
+
+        unicode_doc = "<!DOCTYPE html>\n{}".format(doc.unicode(indent=2))
+
+        # Fix encoding issues, e.g. with surrogates
+        unicode_doc = unicode_doc.encode("utf-8", errors="xmlcharrefreplace")
+        return unicode_doc.decode("utf-8")
+
+    def _generate_environment(self, config):
+        if not hasattr(config, "_metadata") or config._metadata is None:
+            return []
+
+        metadata = config._metadata
+        environment = [html.h2("Environment")]
+        rows = []
+
+        keys = [k for k in metadata.keys()]
+        if not isinstance(metadata, OrderedDict):
+            keys.sort()
+
+        for key in keys:
+            value = metadata[key]
+            if self._is_redactable_environment_variable(key, config):
+                black_box_ascii_value = 0x2593
+                value = "".join(chr(black_box_ascii_value) for char in str(value))
+
+            if isinstance(value, str) and value.startswith("http"):
+                value = html.a(value, href=value, target="_blank")
+            elif isinstance(value, (list, tuple, set)):
+                value = ", ".join(str(i) for i in sorted(map(str, value)))
+            elif isinstance(value, dict):
+                sorted_dict = {k: value[k] for k in sorted(value)}
```

Comment:
```suggestion
                value = json.dumps(value,  sort_keys=True)
```

a future addon might be to have a interactive intended view of the data that can expand/fold